src/app/website/posts/zkp.markdown

```diff
@@ -0,0 +1,81 @@
+---
+title: Zero-knowledge Proofs&#58; An Intuitive Explanation
+date: 2019-05-13
+author: Vanishree Rao
+---
+
+Zero-knowledge proofs (ZKPs) are a powerful cryptography primitive that enables you to prove that you have a secret, without revealing it to anyone.
+ If you are hearing about ZKPs for the first time, you are likely to say "Hah! That sounds impossible." Read on to get an intuitive understanding of what they are. But first, some background. ZKPs were [invented](http://people.csail.mit.edu/silvio/Selected%20Scientific%20Papers/Proof%20Systems/The_Knowledge_Complexity_Of_Interactive_Proof_Systems.pdf) by Shafi Goldwasser, Silvio Micali, and Charles Rackoff in 1985. Ever since, ZKPs have been one of the most active areas of research in Cryptography. Moreover, recently, they are enjoying significant impact on real-world applications, specifically on blockchain technologies. [Zcash](https://z.cash/technology/), a pioneering blockchain project, employed ZKPs to achieve anonymity in financial transactions. At O(1)Labs, we are building CODA, the first succinct blockchain, using ZKPs. No matter how many transactions are recorded on the blockchain, the blockchain remains at most the size of a few tweets. 
+
+
+
+## What are Zero-knowledge Proofs?
+
+Now let’s consider a puzzle and see how you could convince anybody that you know a solution without revealing it. 
+
+
+### A Puzzle called 3-Coloring. 
+
+The 3-coloring puzzle can be described as follows. You are given a graph of nodes and edges (like in the figure below). The task is to find a "3-coloring" to the graph, which is a coloring of the nodes with three different colors in such a way that no two adjacent nodes have the same color. 
+
+![An uncolored graph.](/static/blog/zkp/uncolored-graph.png)
+
+![A colored graph.](/static/blog/zkp/colored-graph.png)
+
+### A Zero-knowledge Proof for the 3-coloring Puzzle. 
+
+We will construct a ZKP protocol for the 3-coloring puzzle. Before that, let’s quickly recall the two properties we are looking for in the protocol. 
+
+A ZKP protocol between you and someone -- call her  Verifier, must satisfy the following properties: 
+
+* If you are cheating (i.e., if you do not know a 3-coloring), then Verifier should be able to catch you -- this property is called  _soundness_
+
+* Verifier should not learn anything about the 3-coloring -- this property is called  _zero knowledgeness_
+
+Now, let's try to construct a ZKP protocol. Imagine the graph drawn on the floor in a closed space. Per the 3-coloring you know, you would place the corresponding colored balls on the nodes. Then you would completely cover the balls with inverted opaque bowls. 
+Then, Verifier comes and points at an edge of her choice. You would lift up the two bowls on either side of the edge. Verifier verifies that the balls revealed are of different colors. If they are not, you are caught cheating. 
```

Comment:
Maybe explicitly separate the proving and verification steps? Something like:

Now, let's try to construct a ZKP protocol. We'll separate the protocol up into two phases, proving and verifying.

1. **Proving:** Imagine the graph drawn on the floor in a closed space. Per the 3-coloring you know, you would place the corresponding colored balls on the nodes. Then you would completely cover the balls with inverted opaque bowls. 
2. **Verifying:** Then, Verifier comes and points at an edge of her choice. You would lift up the two bowls on either side of the edge. Verifier verifies that the balls revealed are of different colors. If they are not, you are caught cheating.

Reply:
Done.